src/autodiff.jl

```diff
@@ -0,0 +1,144 @@
+function partial_contraction(ψ::MPS, ϕ::MPS)
+  T = ITensor(1)
+  for n in 1:length(ψ)
+    T = T * ψ[n] * ϕ[n]
+  end
+  return T
+end
+
+function inner(ϕ::ITensor, U::Vector{ITensor}, ψ::ITensor)
+  Uψ = ψ
+  for u in U
+    s = commoninds(u, Uψ)
+    s′ = s'
+    Uψ = replaceinds(u * Uψ, s′ => s)
+  end
+  return (dag(ϕ) * Uψ)[]
+end
+
+
+
+function inner(ϕ::MPS, U::Vector{ITensor}, ψ::MPS; kwargs...)
+  Uψ = runcircuit(ψ, U; kwargs...)
+  return ITensors.inner(ϕ, Uψ)
+end
+
+inner(ϕ::MPS, circuit::Vector{<:Tuple}, ψ::MPS; kwargs...) = 
+  inner(ϕ, buildcircuit(ψ, circuit), ψ; kwargs...)
+
+
+function rrule(::typeof(inner), ϕ::MPS, U::Vector{ITensor}, ψ::MPS; cache::Bool = true, kwargs...)
+  Udag = reverse([dag(swapprime(u, 0=>1)) for u in U])
+  if cache
+    ξ⃗ = MPS[ϕ]
+    for udag in Udag
+      ξ⃗  = vcat(ξ⃗, apply(udag, ξ⃗[end]; move_sites_back = true, kwargs...))
+    end
+    ξ⃗l = reverse(ξ⃗)
+    y = inner(ξ⃗l[1], ψ)
+  else
+    ξl = runcircuit(ϕ, Udag; kwargs...)
+    y = inner(ξl, ψ)
+  end
+
+  function inner_pullback(ȳ)
+    ∇⃗ = ITensor[]
+    ξr = copy(ψ)
+    for (i,u) in enumerate(U)
+      if cache
+        ξl = ξ⃗l[i+1]
+      else
+        ξl = apply(u, ξl; move_sites_back = true, kwargs...)
+      end
+      ξl = prime(ξl, inds(u, plev = 0))
+      ∇⃗ = vcat(∇⃗, partial_contraction(ξl, dag(ξr)))
+      noprime!(ξl)
+      ξr = apply(u, ξr; move_sites_back = true, kwargs...)
+    end
+    return (NoTangent(), NoTangent(), ȳ .* ∇⃗, NoTangent())
+  end
+  return y, inner_pullback
+end
+
+
+function inner(O::MPO, U::Vector{ITensor}, ψ::MPS; kwargs...)
```

Comment:
I think this is a confusing use of the name `inner`, what if we called it `rayleigh_quotient`?

Reply:
Yeah I was going to bring this up with you. Sounds good.

Reply:
The only "issue" I see is that it may be confusing that technically the Rayleigh quotient should be `<Uψ|O|Uψ>/<Uψ|Uψ>`, but in general we don't really need to divide by the normalization. We could have a keyword argument `normalize=false` which optionally divides by `<Uψ|Uψ>` which is off by default. We don't need that now but I'm just bringing it up as something to think about.

In addition, the circuit may or may not be unitary, in which case the normalization `<Uψ|Uψ>` may or may not reduce to `<ψ|ψ>`, which we could also control with an `isunitary` flag or something like that. Again, not something we have to do now, just anticipating future design issues.

Reply:
Good point, yeah let's keep this in mind. This will end up being moved anyway a little down the road.